### PR TITLE
[14.0] [IMP] account_statement_import_online[_ponto] improvements.

### PR DESCRIPTION
**Pull request by NL66278**
1. Find partner if there already is a bank account with the right IBAN;
2. Store (and display) raw import data to help in problem determination;
3. Make it easy to extend the parsing of import data;
4. Apply some clean coding principles.

### CODE REVIEW

File: account_statement_import_online/models/account_bank_statement_line.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+from odoo import fields, models
+
+
+class AccountBankStatementLine(models.Model):
+    """Add raw data to statement line, to solve import issues."""
+
+    _inherit = "account.bank.statement.line"
+
+    raw_data = fields.Text(
```
Comment (pedrobaeza) [inline account_statement_import_online/models/account_bank_statement_line.py]:
```suggestion
    online_raw_data = fields.Text(
```
?

File: account_statement_import_online/models/online_bank_statement_provider.py [hunk 2]
```diff
@@ -365,6 +363,14 @@ def _sanitize_bank_account_number(self, bank_account_number):
         self.ensure_one()
         return sanitize_account_number(bank_account_number)
 
+    def _update_partner_from_account_number(self, line_values):
+        partner_bank = self.env["res.partner.bank"].search(
+            [("acc_number", "=", line_values["account_number"])], limit=1
```
Comment (pedrobaeza) [inline account_statement_import_online/models/online_bank_statement_provider.py]: No `company_id` restriction?

File: account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py [hunk 3]
```diff
@@ -156,34 +156,32 @@ def _ponto_get_transaction(self, account_id, date_since, date_until):
...
-                        if not page_next or (page_next and not latest_identifier):
-                            latest_identifier = current_transactions[0].get("id")
-                        transaction_lines.extend(current_transactions)
-            else:
+            if response.status_code != 200:
                 raise UserError(
                     _("Error during get transaction.\n\n%s \n\n %s")
                     % (response.status_code, response.text)
                 )
+            if params.get("before"):
```
Comment (pedrobaeza) [inline account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py:164]: Why changing this order? Please try to minimize diff

File: account_statement_import_online/models/account_bank_statement_line.py (hunk 1 above)
Comment (NL66278) [reply account_statement_import_online/models/account_bank_statement_line.py]: I will change the name

File: account_statement_import_online/models/online_bank_statement_provider.py (hunk 2 above)
Comment (NL66278) [reply account_statement_import_online/models/online_bank_statement_provider.py]: I took this from the not online account_statement_import. res.partner.bank statements with company_id filled are own accounts, not partner accounts. Maybe I should add that company_id should be False.

File: account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py (hunk 3 above)
Comment (NL66278) [reply account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py:164]: I thought it was cleaner, saving a lot of indented code and getting rid of an `else` branch. But if you still object I will change it back.

File: account_statement_import_online/models/online_bank_statement_provider.py (hunk 2 above)
Comment (pedrobaeza) [reply account_statement_import_online/models/online_bank_statement_provider.py]: Can you have the same bank account number in several companies? If so, then you should check by company_id = company of the statement or company_id = False.

File: account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py (hunk 3 above)
Comment (pedrobaeza) [reply account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py:164]: Although I can agree on the principle, this makes more difficult to review the improvements. Maybe you can add an initial commit doing this refactoring, and another with the improvements.

File: account_statement_import_online/models/online_bank_statement_provider.py (hunk 2 above)
Comment (NL66278) [reply account_statement_import_online/models/online_bank_statement_provider.py]: O wait, that was in in old version of Odoo.

Comment (NL66278) [reply account_statement_import_online/models/online_bank_statement_provider.py]: Still, I do not see a need to check for company. You can have a partner without a company_id, meaning valid everywhere, and then when adding a bank_account, the res.partner.bank record gets the company_id of the logged in user. I think that is illogical, and unneeded, as it is still not a required field.

Comment (pedrobaeza) [reply account_statement_import_online/models/online_bank_statement_provider.py]: I insist: the problem is that you have the same bank account number in several companies, and with this, you select the incorrect bank account belonging to other company.

**Comment by NL66278** [comment]
@pedrobaeza I think I attended to all your comments now.

**Review by alexis-via — APPROVED** [review]
code review, no test

**Review by rlizana — APPROVED** [review]
Code review, @pedrobaeza comments is satisfied, not test!

**Pull request merged**